README.md
## Anonsurf

A tool by [ParrotSec](https://www.parrotsec.org/docs/anonsurf.html) written in Nim as an anonymous mode wrapper to force connections through [Tor](https://www.torproject.org/)

This fork changes the debian style service and directory structure, able to run on Archlinux which assumes
networkmanager, tor, iptables and dnsmasq are installed

### Tor Bridges
- List of default bridges are at https://gitlab.torproject.org/tpo/anti-censorship/team/-/wikis/Default-Bridges
### Thank to
- Ranjeetsih, who wrote example of sys-tray with GTK at https://www.codeproject.com/Articles/27142/Minimize-to-tray-with-GTK
- Vivek Gite, who wrote tutorial of hook script for dhclient https://www.cyberciti.biz/faq/dhclient-etcresolvconf-hooks/

debian/changelog
anonsurf (4.1.1+parrot2) lory; urgency=medium

  * Rebuild for Parrot 6.

 -- Lorenzo "Palinuro" Faletra <palinuro@parrotsec.org>  Tue, 19 Dec 2023 09:54:15 +0100

anonsurf (4.1.1+parrot1) parrot-lts; urgency=medium

  * Fix version name for stable release

 -- dmknght <dmknght@parrot>  Thu, 02 Feb 2023 15:59:34 +0700

anonsurf (4.1.0+parrot0) parrot-lts; urgency=medium

  [ Nong Hoang Tu ]
  * Fixed anonsurf-gtk iat-mode regex
  * Fixed a critical bug when generate bridge mode in torrc
  * Change anonsurf-gtk to use Application. Fixed multiple systray
  * Move onionshares to suggests
  * Add Vala source code as WIP gui
 -- dmknght <dmknght@parrot>  Mon, 23 Jan 2023 14:16:02 +0700

anonsurf (4.0.0+parrot1) parrot-updates; urgency=medium

  * 4.0 release

 -- Nong Hoang Tu <dmknght@parrotsec.org>  Sun, 21 Aug 2022 23:27:49 +0700

anonsurf (3.3.3+parrot1) parrot-updates; urgency=medium

  [ Nong Hoang Tu ]
  * Move bottom bar to header bar. Close #7
  * Use EventBox to switch from Main to Details widget
  * Remove unused objects
  * Switch to main from Details. Close #8
  * Add descriptions about Gintro version
  * Change version to 4.0.0 pre
  * Check if runResolvconf exists
  * Create type MessageCallback for callback function. It looks better than "proc" only
  * Remove old fixme which was fixed
  * Reverse the proc type for callback, which doesn't work for macro
  * Allow function check ip to be used by right click menu
  * Quit now uses the same function with core
  * Make restore main widget by left click work. Close #12
  * Remove unused comments
  * Add submenu with start / stop / restart/ status of anonsurf
  * Rewrite module main widget for better code structure
  * Change module name to ansurf_gtk_objects instead of ansurf_objects, which less confussing
  * New code structure for details widget
  * Optmize spacing between frames
  * All dialogs are no longer in folder "widgets"
  * Use same callback for menu and gui. Close #11
  * Add thank to author who wrote tutorial of systray + gtk
  * Move status to main menu
  * Swap position of about and settings
  * Status backend now shows AnonSurf not running using callback
  * Connect systray's status with backend
  * Change name "AnonSurf GUI" to "AnonSurf"
  * Create function init_main_window for window app settings
  * Merge code stack with the create Layout
  * Remove boxMainWindow show all because mainboard did this
  * Move back systray icon to main. systray icon doesn't appear when it's in child function
  * Smaller refresh boot by move update boot to other functions
  * Improve function name
  * Move all condition blocks to small functions
  * Move functions to other module
  * Move some functions to other module
  * Improve code spacing
  * Create function to update ports and services status
  * Remove trail characters
  * Remove unused libs
  * Fix wrong tag close
  * Make condition be more simple
  * Improve function name
  * Move handle refresh to refresher module
  * Imrprove name and move all refresh code to module fresher
  * Improve comment for refresh
  * Move code to make stack to other small function
  * Create a object for right click menu Item
  * remove unused comments
  * Move all backend code of menuitem to other module. Complete features
  * Update notification in makefile
  * Change callback args to enum
  * Remove comment
  * code clean
  * Improve object types
  * Code clean and close #6
  * Move module kill_apps, improve some module names
  * Add more options to settings
  * Add tooltip for bridge
  * Add option to block inbound traffic. force this feature to optional
  * New code structure
  * Move some old functions to new code
  * Add some more tmp functions
  * shorter condition by using function
  * Add button bypass network firewall to complete later
  * improve var name
  * add some hardcoded print
  * Move functions to small modules
  * More small functions
  * sightly change function structure
  * Improve condition to check dhclient's files
  * More print
  * Some more template functions
  * Improve code format
  * some more extra tmp tools
  * Add a proc to handle all dns addr
  * Filter valid addr
  * basic handle create symlink with resolvconf
  * Validate dns addr
  * Complete more functions
  * New way to handle dns addr from user
  * Remove todo
  * Complete mix with dhcp when tool can't find dhcp addr
  * Use deduplicate in write_dns_addr_to_file so it will be called by all functions
  * Make parse dhclient dns be global
  * Complete mix addr
  * Check if dns is empty before write
  * Improve dhclient code by calling pre-write functions
  * Move backup functions to util and create function to check if anonsurf is running
  * change function name to print_file_static
  * Add function to check localhost and the other features of status
  * Remove old dnstool
  * Add new dnstool to makefile build. close #9
  * Add resolvconf to suggest list
  * Improve error msg when option is invalid
  * Move check_system_dns_is_static to utils
  * Fix grammar
  * Fix wrong file to be called to parse dns addresses
  * Shorter descriptions
  * Add print function to print banner of addresses
  * Check if current system is having only localhost. Do not create backup if true
  * Add tmp function to create hook script
  * Improve code syntax
  * Create a function to do destroy dialog
  * Add some buttons to dialogs
  * Improve create backup logic check
  * Prepare objects for setting dialog
  * Move all option objects to new module
  * No longer use multiple spacing
  * Correct the symlink info to "resolvconf settings"
  * Correct the symlink info to "resolvconf settings"
  * reverse wrong commit that test text align
  * Add some basic tmp functions to handle options
  * Add a pipe object
  * Add one other function
  * Add short comment with video tutorial
  * Add more tmp functions for workflow
  * Create function to get value and set value for Config object
  * Change value type to string

src/newGUISource.nim
import gintro / [gtk, glib, gobject]#, notify, vte]
import utils / [status, dnsutils, help]

type
  rObject = ref object
    rLabelFastStatus: Label
    rLabelAnonStatus: Label
    rLabelTorStatus: Label
    rLabelDNSStatus: Label
    rLabelStatusBoot: Label
    rImgStatus: Image
    rBtnNyx: Button
    rBtnStart: Button
    rBtnStartBridge: Button
    rBtnStop: Button
    rBtnRestart: Button
    rBtnChangeID: Button
    rBtnSetBoot: Button

const # TODO change icon path here
  imageFailed = "/home/dmknght/Parrot_Projects/anonsurf/icons/50px/anon_50px_failed.png"
  imageUnprotected = "/home/dmknght/Parrot_Projects/anonsurf/icons/50px/anon_50px_unprotected.png"
  imageProtected = "/home/dmknght/Parrot_Projects/anonsurf/icons/50px/anon_50px_protected.png"


proc onClickDashboard(b: Button, s: Stack) =
  #[
    init action for back button and button in Dashboard
  ]#
  if b.label == "Back" or b.label == "":
    s.setVisibleChildName("main")
  else:
    if b.label == "AnonSurf":
      s.setVisibleChildName("anonsurf")
    elif b.label == "Set Boot":
      s.setVisibleChildName("boot")
    elif b.label == "More Tools":
      s.setVisibleChildName("tools")
    elif b.label == "Details":
      s.setVisibleChildName("details")


proc onClickExit(b: Button) =
  mainQuit()


proc refreshStatus(args: rObject): bool =
  #[
    Update current status everytime
  ]#
  let
    basicStatus = getSurfStatus()
    dnsStatus = dnsStatusCheck()
  
  case dnsStatus
  of 0:
    args.rLabelDNSStatus.text = "Under Tor Tunnel"
  of -1:
    args.rLabelDNSStatus.text = "Localhost (error)"
  of -2:
    args.rLabelDNSStatus.text = "/etc/resolv.conf not found"
  of -3:
    args.rLabelDNSStatus.text = "/etc/resolv.conf is empty"
  of 11:
    args.rLabelDNSStatus.text = "Static: OpenNIC DNS"
  of 12:
    args.rLabelDNSStatus.text = "Static: custom DNS"
  of 13:
    args.rLabelDNSStatus.text = "Static: OpenNIC + Custom DNS"
  of 20:
    args.rLabelDNSStatus.text = "Dynamic: DHCP"
  of 21:
    args.rLabelDNSStatus.text = "Dynamic: DHCP with OpenNIC DNS"
  of 22:
    args.rLabelDNSStatus.text = "Dynamic: DHCP with Custom DNS"
  of 23:
    args.rLabelDNSStatus.text = "Dynamic: DHCP with OpenNIC + Custom DNS"
  else:
    discard

  #[
    Update boot's status by the label
  ]#
  if basicStatus.isAnonSurfBoot:
    args.rLabelStatusBoot.text = "It is enabled"
    args.rBtnSetBoot.setLabel("Disable")
  else:
    args.rLabelStatusBoot.text = "It is not enabled"
    args.rBtnSetBoot.setLabel("Enable")

  #[
    Update Tor's status by the label
  ]#
  case basicStatus.isTorService
  of 1: # Tor is running
    args.rLabelTorStatus.text = "Tor is running"
  of 0: # Tor is not running
    args.rLabelTorStatus.text = "Tor is not running"
  else: # Tor failed to run
    args.rLabelTorStatus.text = "Tor start failed (error)"

  #[
    Update other labels and buttons base on AnonSurf status
  ]#
  case basicStatus.isAnonSurfService
  of 1: # AnonSurf is running
    # Show status of labels
    args.rLabelAnonStatus.text = "AnonSurf is running"
    # Disable buttons that can start AnonSurf
    args.rBtnStart.setSensitive(false)
    args.rBtnStartBridge.setSensitive(false)
    # Enable buttons that can change AnonSurf running state
    args.rBtnRestart.setSensitive(true)
    args.rBtnStop.setSensitive(true)
    args.rBtnChangeID.setSensitive(true)
    # Enable nyx to check AnonSurf status
    args.rBtnNyx.setSensitive(true)
    # Check fast status
    if basicStatus.isTorService == 1:
      args.rImgStatus.setFromFile(imageProtected)
      args.rLabelFastStatus.text = "AnonSurf is running"
    elif basicStatus.isTorService == 0:
      args.rImgStatus.setFromFile(imageFailed)
      args.rLabelFastStatus.text = "AnonSurf is running but Tor isn't"
    else:
      args.rImgStatus.setFromFile(imageFailed)
      args.rLabelFastStatus.text = "AnonSurf is running but Tor failed"
  of 0: # AnonSurf is not running
    # Update image
    args.rImgStatus.setFromFile(imageUnprotected)
    # Show status of label
    args.rLabelAnonStatus.text = "AnonSurf is not running"
    # Enable buttons that can start AnonSurf
    args.rBtnStart.setSensitive(true)
    args.rBtnStartBridge.setSensitive(true)
    # Disable buttons that can change AnonSurf running state
    args.rBtnRestart.setSensitive(false)
    args.rBtnStop.setSensitive(false)
    args.rBtnChangeID.setSensitive(false)
    # Disable nyx button
    args.rBtnNyx.setSensitive(false)
    # Update fast status
    args.rLabelFastStatus.text = "AnonSurf is not running"
  else:
    # Update image
    args.rImgStatus.setFromFile(imageFailed)
    # Show status of label
    args.rLabelAnonStatus.text = "AnonSurf failed to run (error)"
    # Disable buttons can start AnonSurf
    args.rBtnStart.setSensitive(false)
    args.rBtnStartBridge.setSensitive(false)
    # Disable restart button because it is having error
    args.rBtnRestart.setSensitive(false)
    args.rBtnChangeID.setSensitive(false)
    # Enable stop button
    args.rBtnStop.setSensitive(true)
    # Disable nyx button
    args.rBtnNyx.setSensitive(false)
    # Update fast status
    args.rLabelFastStatus.text = "AnonSurf failed to run"

  return SOURCE_CONTINUE


proc createArea(boxMainWindow: Box) =
  #[
    Create a Kaspersky-like UI or XFCE4 control Pannel like UI
    When we click on a Button, main widget changes to button's display
    Use the Button name (TODO change button object) as same as child name
  ]#
  let
    #[
      Box dashboard is the main object when we open new AnonSurf.
      It should have:
        1. First row: Status of current AnonSurf
        2. Second row: All buttons to switch to other widgets / pages
    ]#
    boxDashboard = newBox(Orientation.vertical, 3)
    #[
      Main stack is the "page manager" for our job
      It manages all page as child object
    ]#
    mainStack = newstack()

  #[  Draw first row
  First column: Icon of AnonSurf. Color as current status
    1. Error: Red
    2. Not connected: yellow
    3. Connected, no error: green
  Second column:
    Show current status of AnonSurf fast status like actived / inactived
    Show details button
  Detail:
    When we click on it, it will show something like:
      Tor: is actived
      AnonSurf: is actived
      DNS under tor
      Nyx in vte terminal
      A button that can change id
      Check IP
  ]#
  let
    # TODO Use icon here
    boxStatus = newBox(Orientation.horizontal, 3)
    boxStatusIcon = newBox(Orientation.vertical, 3)
    boxStatusDetails = newBox(Orientation.vertical, 3)
    labelStatus = newLabel("Your connection isn't protected")
    btnDetails = newButton("Details")

  boxStatus.setSizeRequest(300, 70)
  boxStatusDetails.add(labelStatus)
  # boxStatus.

  boxStatusDetails.add(btnDetails)

  let # TODO change icon path here
    imgStatus = newImageFromFile(imageUnprotected)

  boxStatusIcon.add(imgStatus)

  boxStatus.packStart(boxStatusIcon, true, true, 3)
  boxStatus.packEnd(boxStatusDetails, false, true, 3)
  
  boxDashboard.add(boxStatus)

  #[
    Draw the second row
    It has button: AnonSurf, Boot, Extra.
    Each button uses to switch from main widget (dashboard) to its widget
  ]#
  let
    btnAnon = newButton("AnonSurf")
    btnBoot = newButton("Set Boot")
    btnExtra = newButton("More Tools")
    boxMainButtons = newBox(Orientation.horizontal, 3)

  btnAnon.connect("clicked", onClickDashboard, mainStack)
  btnAnon.setSizeRequest(80, 80)
  boxMainButtons.packStart(btnAnon, false, true, 3)

  btnBoot.connect("clicked", onClickDashboard, mainStack)
  btnBoot.setSizeRequest(80, 80)
  boxMainButtons.add(btnBoot)

  btnExtra.connect("clicked", onClickDashboard, mainStack)
  btnExtra.setSizeRequest(80, 80)
  boxMainButtons.packEnd(btnExtra, false, true, 3)
  boxMainButtons.show()
  boxDashboard.add(boxMainButtons)

  let
    boxBottomButtons = newBox(Orientation.horizontal, 3)
    btnMainHelp = newButton("About")
    btnExit = newButton("Exit")
    imgMainHelp = newImageFromIconName("help-about", 3)
    imgMainExit = newImageFromIconName("exit", 3)

  btnExit.connect("clicked", onClickExit)
  btnExit.setImage(imgMainExit)
  boxBottomButtons.packStart(btnExit, false, true, 3)
  btnMainHelp.connect("clicked", onClickMainHelp)
  btnMainHelp.setImage(imgMainHelp)
  boxBottomButtons.packEnd(btnMainHelp, false, true, 3)

  boxDashboard.packEnd(boxBottomButtons, false, true, 3)


  ## End of draw second row
  boxDashboard.showAll()
  mainStack.addNamed(boxDashboard, "main")
  # End of draw dashboard

  #[
    Draw AnonSurf board
    This board is used to use main AnonSurf procedures
    It has start / start bridge / stop / restart
  ]#
  let
    btnStart = newButton("Start")
    btnStartBridge = newButton("Start Bridge")
    labelAnonStart = newLabel("Start Normal")
    labelAnonStartBridge = newLabel("Start with Obfs4 Bridge")
    boxAnonStart = newBox(Orientation.vertical, 3)
    boxAnonStartBridge = newBox(Orientation.vertical, 3)

  boxAnonStart.add(labelAnonStart)
  boxAnonStart.add(btnStart)

  boxAnonStartBridge.add(labelAnonStartBridge)
  boxAnonStartBridge.add(btnStartBridge)
  
  let
    boxAnonStartArea = newBox(Orientation.horizontal, 3)
  boxAnonStartArea.add(boxAnonStart)
  boxAnonStartArea.packEnd(boxAnonStartBridge, false, true, 3)

  let
    btnRestart = newButton("Restart")
    labelRestart = newLabel("Restart AnonSurf")
    boxAnonRestart = newBox(Orientation.vertical, 3)
    
  boxAnonRestart.add(labelRestart)
  boxAnonRestart.add(btnRestart)

  let
    btnStop = newButton("Stop")
    labelStop = newLabel("Stop AnonSurf")
    boxAnonStop = newBox(Orientation.vertical, 3)

  boxAnonStop.add(labelStop)
  boxAnonStop.add(btnStop)

  let
    boxAnonBottomButtons = newBox(Orientation.horizontal, 3)
    btnAnonBack = newButton("")
    btnAnonHelp = newButton("Help")
    imgAnonHelp = newImageFromIconName("help-about", 3)
    imgAnonBack = newImageFromIconName("back", 3)
  
  btnAnonBack.connect("clicked", onClickDashboard, mainStack)
  btnAnonBack.setImage(imgAnonBack)
  btnAnonHelp.setImage(imgAnonHelp)
  boxAnonBottomButtons.packStart(btnAnonBack, false, true, 3)
  boxAnonBottomButtons.packEnd(btnAnonHelp, false, true, 3)

  let
    boxAnonArea = newBox(Orientation.vertical, 3)

  boxAnonArea.add(boxAnonStartArea)
  boxAnonArea.add(boxAnonRestart)
  boxAnonArea.add(boxAnonStop)
  boxAnonArea.packEnd(boxAnonBottomButtons, false, true, 3)

  boxAnonArea.show()
  
  mainStack.addNamed(boxAnonArea, "anonsurf")
  # End of AnonSurf board
  
  #[
    Draw Boot option board
    The label will be updated by using refresh
    btnBoot will be changed by current boot status
  ]#
  let
    boxBoot = newBox(Orientation.vertical, 3)
    labelBootStatus = newLabel("disabled")
    btnBootSwitch = newButton("Enable")
    btnBootBack = newButton("Back")

  boxBoot.add(labelBootStatus)
  boxBoot.add(btnBootSwitch)
  btnBootBack.connect("clicked", onClickDashboard, mainStack)
  boxBoot.add(btnBootBack)
  boxBoot.show()

  mainStack.addNamed(boxBoot, "boot")
  # End of boot dashboard

  #[
      Draw Extra option dashboard
      It will have DNS tool in GUI and MAC changer
  ]#
  let
    boxExtra = newBox(Orientation.vertical, 3)
    btnDNS = newButton("dns")
    # btnMAC = newButton("mac")
    btnToolsBack = newButton("Back")

  boxExtra.add(btnDNS)
  # boxExtra.add(btnMAC)
  btnToolsBack.connect("clicked", onClickDashboard, mainStack)
  boxExtra.add(btnToolsBack)
  boxExtra.show()

  mainStack.addNamed(boxExtra, "tools")
  # End of extra dashboard

  #[
    Show a new dialog about all details
    Services:
      Tor: is actived
      AnonSurf: is actived
    DNS:
      DNS under tor
    Buttons:
      A button that can change id
      Check IP
    Monitor:
      Nyx in vte terminal - click button
  ]#
  #[
    Box 1: vertical: AnonSurf, Tor, DNS details
    Box 2: Horizontal: Buttons: changeID, checkIP
  ]#

  #[
    Add status
  ]#
  let
    boxStatusServices = newBox(Orientation.vertical, 3)
    labelAnonStatus = newLabel("AnonSurf: Not running")
    labelTorStatus = newLabel("Tor: Not running")
    labelDNSStatus = newLabel("DNS: DNS status")

  boxStatusServices.add(labelAnonStatus)
  boxStatusServices.add(labelTorStatus)
  boxStatusServices.add(labelDNSStatus)

  #[
    Add buttons
  ]#
  let
    boxStatusButtons = newBox(Orientation.horizontal, 3)
    btnChangeID = newButton("Change ID")
    btnCheckIP = newButton("Check IP")
    btnNyx = newButton("Nyx")

  btnDetails.connect("clicked", onClickDashboard, mainStack)
  
  boxStatusButtons.add(btnChangeID)
  boxStatusButtons.add(btnCheckIP)
  boxStatusButtons.add(btnNyx)

  let
    boxDetailsBottomButtons  = newBox(Orientation.horizontal, 3)
    btnDetailsBack = newButton("")
    imgDetailsBack = newImageFromIconName("back", 3)

  btnDetailsBack.connect("clicked", onClickDashboard, mainStack)
  btnDetailsBack.setImage(imgDetailsBack)
  boxDetailsBottomButtons.add(btnDetailsBack)

  let
    boxDetails = newBox(Orientation.vertical, 3)

  boxDetails.add(boxStatusServices)
  boxDetails.add(boxStatusButtons)
  boxDetails.packEnd(boxDetailsBottomButtons, false, true, 3)

  boxDetails.showAll()
  
  mainStack.addNamed(boxDetails, "details")

  # End of full details dashboard

  var
    refreshObjs: rObject

  refreshObjs = rObject(
    rLabelFastStatus: labelStatus,
    rLabelAnonStatus: labelAnonStatus,
    rLabelTorStatus: labelTorStatus,
    rLabelDNSStatus: labelDNSStatus,
    rLabelStatusBoot: labelBootStatus,
    rImgStatus: imgStatus,
    rBtnNyx: btnNyx, 
    rBtnStart: btnStart,
    rBtnStartBridge: btnStartBridge,
    rBtnStop: btnStop,
    rBtnRestart: btnRestart,
    rBtnChangeID: btnChangeID,
    rBtnSetBoot: btnBootSwitch,
  )

  discard timeoutAdd(200, refreshStatus, refreshObjs)
  
  # Add everything to window
  boxMainWindow.add(mainStack)
  boxMainWindow.showAll()


proc stop(w: Window) =
  mainQuit()

proc main =
  #[
    Create new window
  ]#
  gtk.init()
  let
    mainBoard = newWindow()
    boxMainWindow = newBox(Orientation.vertical, 3)
  
  mainBoard.title = "AnonSurf GUI"
  discard mainBoard.setIconFromFile("/usr/share/icons/anonsurf.png")

  createArea(boxMainWindow)

  mainBoard.add(boxMainWindow)
  mainBoard.setBorderWidth(3)

  mainBoard.show()
  mainBoard.connect("destroy", stop)
  gtk.main()

main()
